scripts/thrift/client.py

```diff
@@ -0,0 +1,142 @@
+#
+# -------------------------------------------------------------------------
+"""
+This is a simple example how we can write our own Python client for a
+CodeChecker server.
+
+For available API functions see the .thrift files:
+  https://github.com/Ericsson/codechecker/tree/master/web/api
+"""
+
+import sys
+
+from typing import Optional
+
+try:
+    # pylint: disable=no-name-in-module
+    from thrift2.transport import THttpClient
+    from thrift2.protocol import TJSONProtocol
+    from thrift2.Thrift import TApplicationException
+except:
+    print("'thrift' package (https://pypi.org/project/thrift/) is not "
+          "available in your environment. Please install it before you run "
+          "this script again.")
+    print("> pip3 install thrift==0.13.0")
+    sys.exit(1)
+
+
+try:
+    from codechecker_api.Authentication_v6 import \
+        codeCheckerAuthentication as AuthAPI_v6
+    from codechecker_api.codeCheckerDBAccess_v6 import \
+        codeCheckerDBAccess as ReportAPI_v6
+    from codechecker_api.ProductManagement_v6 import \
+        codeCheckerProductService as ProductAPI_v6
+    from codechecker_api.ServerInfo_v6 import \
+        serverInfoService as ServerInfoAPI_v6
+
+    from codechecker_api_shared.ttypes import RequestFailed
+except:
+    print("'codechecker_api' and 'codechecker_api_shared' must be available "
+          "in your environment to run this script. Please install it before "
+          "you run this script again:")
+    print("  - https://github.com/Ericsson/codechecker/blob/master/web/api/py/codechecker_api/dist/codechecker_api.tar.gz")
+    print("  - https://github.com/Ericsson/codechecker/blob/master/web/api/py/codechecker_api_shared/dist/codechecker_api_shared.tar.gz")
+    sys.exit(1)
+
+
+PROTOCOL = "http"
+HOST = "localhost"
+PORT = 8001
+
+USERNAME = "<myusername>"
+PASSWORD = "<mypassword>"
+
+CLIENT_API = "6.47"
+
+
+def create_client(
+    cls,
+    endpoint: str,
+    product_name: Optional[str] = None,
+    token: Optional[str] = None
+):
+    """ Create a Thrift client. """
+    url = f"{PROTOCOL}://{HOST}:{PORT}/"
+    if product_name:
+        url += f"{product_name}/"
+
+    url += f"v{CLIENT_API}/{endpoint}"
+
+    transport = THttpClient.THttpClient(url)
+    protocol = TJSONProtocol.TJSONProtocol(transport)
+
+    if token:
+        transport.setCustomHeaders({
+            'Cookie': f'__ccPrivilegedAccessToken={token}'})
+
+    return cls.Client(protocol)
+
+
+def main():
+    """ Send multiple Thrift API requests to the server. """
+    # Get server info.
+    cli_server_info = create_client(ServerInfoAPI_v6, "ServerInfo")
+    package_version = cli_server_info.getPackageVersion()
+    print(f"Package version: {package_version}\n")
+
+    # Login to a running CodeChecker server.
+    cli_auth = create_client(AuthAPI_v6, "Authentication")
+
+    try:
+        print(f"Login '{USERNAME}'...")
+        token = cli_auth.performLogin(
```

Comment:
I have a server locally which doesn't need any authentication. I think this is a good thing that login example is here so the user of this script can see its usage. But shouldn't we make it optional? For example if `USERNAME` and `PASSWORD` variables are `None` then `token` should also be `None` so it works without authentication, or something like this.

Reply:
From the server we can ask this information whether it requires authentication or not with the `getAuthParameters` API function. I changed this script to support this use case too what you have mentioned.